pkg/volume/flexvolume/probe.go

```diff
@@ -127,48 +140,87 @@ func (prober *flexVolumeProber) handleWatchEvent(event fsnotify.Event) error {
 	if err != nil {
 		return err
 	}
-
+	parentPathAbs := filepath.Dir(eventPathAbs)
 	pluginDirAbs, err := filepath.Abs(prober.pluginDir)
 	if err != nil {
 		return err
 	}
 
-	// If the Flexvolume plugin directory is removed, need to recreate it
-	// in order to keep it under watch.
-	if eventOpIs(event, fsnotify.Remove) && eventPathAbs == pluginDirAbs {
-		if err := prober.createPluginDir(); err != nil {
+	if eventOpIs(event, fsnotify.Create) || eventOpIs(event, fsnotify.Write) || eventOpIs(event, fsnotify.Rename) {
+		fileinfo, err := prober.fs.Stat(eventPathAbs)
+		if err != nil {
 			return err
 		}
-		if err := prober.addWatchRecursive(pluginDirAbs); err != nil {
-			return err
+		if fileinfo.IsDir() {
+			// dir is created directly in pluginDirAbs
+			if eventOpIs(event, fsnotify.Create) && parentPathAbs == pluginDirAbs {
+				if err := prober.addWatchRecursive(eventPathAbs); err != nil {
+					return err
+				}
+			}
+		} else {
+			if prober.probeAllNeeded {
+				return nil
+			}
+			parentInfo, err := prober.fs.Stat(parentPathAbs)
+			if err != nil {
+				return err
+			}
+			parentBaseName := filepath.Base(parentPathAbs)
+			executableName := filepath.Base(eventPathAbs)
+			parts := strings.Split(parentBaseName, "~")
+			if len(parts) == 1 && executableName != parentBaseName {
+				return nil
+			}
+			if len(parts) == 2 && executableName != parts[1] {
+				return nil
+			}
+			if parentInfo.IsDir() && parentBaseName[0] != '.' {
+				probeEvent := volume.ProbeEvent{
+					Op: volume.AddOrUpdate,
+				}
+				plugin, pluginErr := prober.factory.NewFlexVolumePlugin(prober.pluginDir, parentInfo.Name())
```

Comment:
Plugin should not be initialized inside the filesystem watch handler, for two reasons:

1. The handler has to be very light, and ideally it only provides a status update for the main Probe() goroutine to process. A heavy handler delays the processing of inotify events.
1. Typically multiple inotify events are triggered when a file is created or modified. This will cause a driver to be initialized multiple times.